make block_diag composite compliant

bdhirsh opened the pull request:
The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* #77714
* #77713
* #77132
* __->__ #77716

bdhirsh committed:
make block_diag composite compliant

[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"




[ghstack-poisoned]

bdhirsh commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -719,6 +719,50 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
   return grad_inputs;
 }
 
+std::vector<Tensor> block_diag_backward(const Tensor & grad, const std::vector<std::vector<int64_t>> &sizes, const std::vector<ScalarType> &dtypes) {
```
I mostly got this logic from our existing implementation of `cat_backward` (which knows how to deal with complex)

bdhirsh committed:
Update on "make block_diag composite compliant"




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"




[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -719,6 +719,50 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
+  if (!grad.defined()) {
+    return grad_inputs;
+  }
+  Tensor grad_;
+  bool grad_is_complex = grad.is_complex();
+  if (grad_is_complex) {
+    grad_ = at::real(grad);
+  }
```
What is going on with complex numbers? Is the problem that there is a weird case when we have mixed dtype inputs to block_diag, some of which can be complex, and we need to handle that?

bdhirsh replied:
yeah this is pretty interesting - I got this from the logic for `cat` backward, and it looks this handles exactly that - mixed complex/non-complex inputs. I tried a mixed complex/non-complex input locally and it only started passing when I added this.

bdhirsh replied:
(there also aren't OpInfo tests for the mixed case, so I'll add some)

zou3519 replied:
That makes sense, I will read the formula more closely. I am not sure OpInfos are supposed to test the mixed dtype case (@mruberry ?) because the framework requests a test with a specific dtype. Maybe we just want a manual test somewhere for this.

bdhirsh replied:
Oh I have this implemented by having the sample inputs function check if the passed-in dtype is complex, and if so it spits out an extra sample input for the [only complex], and [mixed complex/noncomplex] case, and the OpInfo suite seemed to pass locally for me. I'll see what happens in CI

bdhirsh commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -719,6 +719,50 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
+      dim0 = shape[0];
+    }
+    grad_inputs[i] = grad_val.slice(0, cur_dim0, cur_dim0 + dim0).slice(1, cur_dim1, cur_dim1 + dim1);
+  }
```
yep you're totally right, this is broken - I was mostly hoping OpInfo tests would point out problems. I'll update and test out a `.backward` call locally

mruberry replied:
Add more sample inputs until they reveal the problem!

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh commented:
@zou3519 I also see some gradcheck tests failing with `RuntimeError: Batching rule not implemented for aten::block_diag. We could not generate a fallback.` (https://github.com/pytorch/pytorch/runs/6496894229?check_suite_focus=true)

How does vmap get used in core testing? (I'm guessing this is the vmap in core, through the `Batched` dispatch key?)

Maybe I just need to opt `block_diag` out of a list somewhere, since it can't use the boxed fallback (it looks like it doesn't support tensorlist inputs)

zou3519 commented:
> @zou3519 I also see some gradcheck tests failing with `RuntimeError: Batching rule not implemented for aten::block_diag. We could not generate a fallback.` (https://github.com/pytorch/pytorch/runs/6496894229?check_suite_focus=true)
> 
> How does vmap get used in core testing? (I'm guessing this is the vmap in core, through the `Batched` dispatch key?)
> 
> Maybe I just need to opt `block_diag` out of a list somewhere, since it can't use the boxed fallback (it looks like it doesn't support tensorlist inputs)

Yes, that's the vmap in core, through the Batched dispatch key. You just need to opt out of it (don't worry about adding a batching rule for block_diag in core, the vmap in core is very behind).

The most likely place where it is being tested is batched gradient computation. You probably want to set at least one of these flags to False in the OpInfo:
![image](https://user-images.githubusercontent.com/5652049/169160242-263e9244-b837-40b8-ae60-991b612db869.png)

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -721,6 +721,58 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
+    }
+    auto& shape = sizes[i];
+    // If input was empty tensor, gradInput should be empty tensor.
+    if (shape == std::vector<int64_t>({0})) {
```
Does the RHS construct a std::vector and throw it away? Because if so, that's not very efficient due to the dynamic allocation

bdhirsh replied:
good point - I'll clean this up (blindly copied from the cat backward kernel)

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -721,6 +721,58 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
+    if (shape.size() == 1) {
+      slice = slice.squeeze(-1);
+    } else if (shape.size() == 0) {
+      slice = slice.squeeze(-1).squeeze(-1);
```
nit: slice.view({}) might be faster, it does one fewer tensor operation

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -721,6 +721,58 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
+    Tensor grad_val;
+    if (!at::isComplexType(dtypes[i]) && grad_is_complex) {
+      // R -> C
+      grad_val = grad_;
```
nit: a better name for `grad_` is "real_view_of_grad" or something

zou3519 reviewed (approved):
LGTM, but tbh I am not sure what is going on with the Lazy shape inference. Is there a test we can add for that?

Also some minor c++ efficiency nits

bdhirsh commented:
@zou3519 I'll clean up the efficiency nits, and I also think I can remove the LTC changes (I forget why I even got the compiler errors that caused them in the first place.. but we shouldn't need them, since LTC can still rely on the decomposition).

I'll merge this - and then once nightlies update and functorch CI starts complaining, I'll merge https://github.com/pytorch/functorch/pull/814. Does that sound ok?

zou3519 commented:
> @zou3519 I'll clean up the efficiency nits, and I also think I can remove the LTC changes (I forget why I even got the compiler errors that caused them in the first place.. but we shouldn't need them, since LTC can still rely on the decomposition).
> 
> I'll merge this - and then once nightlies update and functorch CI starts complaining, I'll merge [pytorch/functorch#814](https://github.com/pytorch/functorch/pull/814). Does that sound ok?

SGTM!

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]

bdhirsh committed:
Update on "make block_diag composite compliant"

The code for `block_diag` isn't composite compliant today - functorch deals with that by registering a special "conditional functionalization" kernel, but I want to kill that here: https://github.com/pytorch/functorch/pull/814.

We can't efficiently make `block_diag` composite compliant though, because it performs `O(num_inputs)` mutations, and converting them all into out-of-place calls would be very inefficient.

Instead, I ended up making `block_diag` `CompositeExplicitAutograd`, and writing a derivative formula for it. That also ended up fixing some OpInfos tests.




[ghstack-poisoned]
